### DISCUSSION TITLE
How to pass args dynamically when writing

### DISCUSSION THREAD

**Discussion by ghost** [wevm/wagmi#905] [Q&A]
I am trying to implement a mint button, this example from docs is basic
https://wagmi.sh/examples/contract-write

my mint function expects some arguments like eg.  _to address. which i want to fetch from use input form . so how can i pass the dynamic value to usePrepareContractWrite .? is it event possible or i should stick to useDeprecatedContractWrite

**Comment by zouhangwithsweet** [comment]
You can write like this

```ts
const { writeAsync: someFn, isLoading } = useContractWrite({
  mode: 'recklesslyUnprepared',
  addressOrName: '0xABCD.....',
  contractInterface: Example_ABI,
  functionName: 'someFn',
  onSuccess: () => {
    notificationSuccess('Write success!')
  },
  onError: () => {
    notificationFailed('Write failed')
  },
})
const doWrite = useCallback(
  async (x: number, y: number) => {
    console.log('someFn', x, y)
    await someFn({
      recklesslySetUnpreparedArgs: [x, y],
    })
  },
  [someFn]
)
```

**Comment by jxom** [comment]
There is now an official example for this! 

https://wagmi.sh/examples/contract-write-dynamic

**Reply by humblecoder** [reply]
The documentation on this is very anemic.  

I have a situation with a very dynamic hook where I need to set nearly everything on-the-fly.  What are the limits here?

What about situations where we want to pass "named" params as args to the function (similar to `react-moralis`)?  Is that a thing?

**Reply by tmm** [reply]
Not familiar with `react-moralis`, but you can pass dynamic params to `write`/`writeAsync` using `useContractWrite` `mode: 'recklesslyUnprepared'`:

```ts
const { write } = useContractWrite({
  address: '0x…',
  abi: […],
  functionName: 'foo',
})

const onClick = () => {
  write({
    // type-safe args if your `abi` is defined inline or const asserted
    // more info here: https://wagmi.sh/react/typescript
    recklesslySetUnpreparedArgs: […],
  })
}

<button onClick={onClick}>write</button>
```

There are a few [UX pitfalls](https://wagmi.sh/react/prepare-hooks#ux-pitfalls-without-prepare-hooks) to be aware of when you do this though.

**Reply by humblecoder** [reply]
Yup, saw that, thanks.  What if I want to configure it all dynamically (including address and abi) at the time of the actual `write` call? 

Also, does this work for `writeAsync` as well?  I'm currently receiving errors that `writeAsync is not a function` which makes me think I'm calling it with incorrect params.  🤔

**Reply by tmm** [reply]
> What if I want to configure it all dynamically (including address and abi) at the time of the actual write call?

You are probably better off using [`useContract`](https://wagmi.sh/react/hooks/useContract) or [`getContract`](https://wagmi.sh/core/actions/getContract) then (although you will lose the reactivity and state tracking that you get with `useContractWrite`).

> Also, does this work for writeAsync as well?

It should! We have a [test case](https://github.com/wagmi-dev/wagmi/blob/main/packages/react/src/hooks/contracts/useContractWrite.test.ts#L475) that uses it.

**Reply by humblecoder** [reply]
First of all, thanks for that. 

I was under the impression that it was a "hooks only" type of situation.  I actually ended up using:
```
import { writeContract, prepareWriteContract } from "@wagmi/core";
```
. . . and everything works as expected.  I can `await prepareWriteContract()` dynamically, and call as needed.  Appreciate the excellent work on the package overall!  👌

**Reply by PaulYoungXTD** [reply]
@tmm, I have similar demand like @humblecoder that the contract address and abi in usePrepareContractWrite() hooks can't be determined after some user actions. The time we know them is just before making the write?.() function call.

Looking forward to adding such dynamic ability to wagmi-react if it is possible.

I also want to say thank you for all the great work done on wagmi. It's awesome!

**Comment by humblecoder** [comment]
Not sure if I should open a separate "discussion", but, as I mentioned earlier, it would be nice if we could simply provide "named params" that get matched/ignored per the passed ABI.  It's certainly possible I misunderstand something about the project, but as it stands, I have to wait until juuuuuuust before the contract call to construct the `args` in a specific order.

I'll see if I can't write something of my own, but this would be a nice to have.

**Reply by tmm** [reply]
> we could simply provide "named params" that get matched/ignored per the passed ABI

If you want to use named params, then you should use a struct at the contract level. The reason `args` is an array is because it represents the arguments that contract functions use.
